src/ledger/recovery.zig

```diff
@@ -0,0 +1,500 @@
+const MerkleProofEntryList = ledger.shred.MerkleProofEntryList;
+const ReedSolomon = ledger.reed_solomon.ReedSolomon;
+const Shred = ledger.shred.Shred;
+const Signature = sig.core.Signature;
+
+const checkedSub = sig.utils.math.checkedSub;
+const makeMerkleTree = ledger.shred.makeMerkleTree;
+const makeMerkleProof = ledger.shred.makeMerkleProof;
+
+const DATA_SHREDS_PER_FEC_BLOCK = ledger.shred.DATA_SHREDS_PER_FEC_BLOCK;
+
+pub const ReedSolomonCache = struct {
+    cache: Cache,
+    const Cache = Lru(
+        .locking,
+        struct { data: usize, parity: usize },
+        ReedSolomon,
+        void,
+        ReedSolomon.deinit,
+    );
+
+    const Self = @This();
+
+    pub fn init(allocator: Allocator) Allocator.Error!Self {
+        return .{ .cache = try Cache.init(allocator, 4 * DATA_SHREDS_PER_FEC_BLOCK) };
+    }
+
+    pub fn deinit(self: *Self) void {
+        self.cache.deinit();
+    }
+
+    /// Caller owns the ReedSolomon. Call `ReedSolomon.deinit` when done.
+    pub fn get(self: *Self, data_shards: usize, parity_shards: usize) !ReedSolomon {
+        if (self.cache.get(.{ .data = data_shards, .parity = parity_shards })) |rs| {
+            var reed_solomon = rs;
+            if (reed_solomon.acquire()) {
+                return reed_solomon;
+            }
+        }
+        var rs = try ReedSolomon.init(self.cache.allocator, data_shards, parity_shards);
+        const acquired = rs.acquire();
+        std.debug.assert(acquired);
+        const old = self.cache.put(.{ .data = data_shards, .parity = parity_shards }, rs);
+        std.debug.assert(old == null);
+        return rs;
+    }
+};
+
+const RecoveryMetadata = struct {
+    common_header: CommonHeader,
+    code_header: CodeShredHeader,
+    chained_merkle_root: ?Hash,
+    retransmitter_signature: ?Signature,
+};
+
+const RecoveryShreds = struct {
+    allocator: Allocator,
+    input_shreds: []?Shred,
+    shards: []?[]const u8,
+    mask: []const bool,
+    num_to_recover: usize,
+
+    /// The shards that are created during reconstruct need to be freed.
+    /// Existing shards are assumed to be owned by another scope.
+    ///
+    /// No shreds are freed, as they are assumed to be owned by another scope.
+    pub fn deinit(self: @This()) void {
+        for (self.mask, self.shards) |was_present, shard| {
+            if (!was_present) {
+                if (shard) |s| self.allocator.free(s);
+            }
+        }
+        self.allocator.free(self.input_shreds);
+        self.allocator.free(self.shards);
+        self.allocator.free(self.mask);
+    }
+
+    // Any shreds that were created during reconstruct are freed.
+    // Existing shards are assumed to be owned by another scope.
+    pub fn deinitShreds(self: @This()) void {
+        for (self.mask, self.input_shreds) |was_present, shred| {
+            if (!was_present) {
+                if (shred) |s| s.deinit();
+            }
+        }
+    }
+};
+
+/// agave: merkle::recover
+pub fn recover(
```

Comment:
can we change this to include permalink + standard analgous format

Reply:
d29e2d6799fd58bae113708d61d31b4eead8f694